docusaurus/docs/Android/04-compose/07-guides/06-implementing-own-capabilities.mdx

```diff
@@ -0,0 +1,233 @@
+            // Default content
+        },
+        commandPopupContent: @Composable (List<Command>) -> Unit = {
+            // Default content
+        },
+        integrations: @Composable RowScope.(MessageComposerState) -> Unit = {
+            // Default content, uses own capabilities to regulate integrations
+        },
+        label: @Composable (MessageComposerState) -> Unit = {
+            // Default content
+        },
+        input: @Composable RowScope.(MessageComposerState) -> Unit = {
+            // Default content
+        },
+        trailingContent: @Composable (MessageComposerState) -> Unit = {
+            // Default content, uses own capabilities to regulate the send button
+        },
+)
+```
+
+`MessageComposer` contains a parameter `messageComposerState: MessageComposerState`. This class holds all the necessary
+state to make `MessageComposer` work, including a set of own capabilities. These are used by the default components
+hosted in the `MessageComposer`.
+
+The file containing `MessageComposer` can be
+found [here](https://github.com/GetStream/stream-chat-android/blob/0f1f54e912baf7a7284d9a4d1435fb6743aa9c78/stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/messages/composer/MessageComposer.kt#L214)
+
+```kotlin
+public data class MessageComposerState(
+        val inputValue: String = "",
+        val attachments: List<Attachment> = emptyList(),
+        val action: MessageAction? = null,
+        val validationErrors: List<ValidationError> = emptyList(),
+        val mentionSuggestions: List<User> = emptyList(),
+        val commandSuggestions: List<Command> = emptyList(),
+        val coolDownTime: Int = 0,
+        val messageMode: MessageMode = MessageMode.Normal,
+        val alsoSendToChannel: Boolean = false,
+        val ownCapabilities: Set<String> = setOf()
+)
+```
+
+`MessageComposer` regulates the following capabilities:
+
+- send-message
+- send-links
+- upload-file
+
+### SelectedMessageMenu
+
+```kotlin
+@Composable
+public fun SelectedMessageMenu(
+        message: Message,
+        messageOptions: List<MessageOptionItemState>,
+        ownCapabilities: Set<String>,
+        onMessageAction: (MessageAction) -> Unit,
+        onShowMoreReactionsSelected: () -> Unit,
+        modifier: Modifier = Modifier,
+        shape: Shape = ChatTheme.shapes.bottomSheet,
+        overlayColor: Color = ChatTheme.colors.overlay,
+        reactionTypes: Map<String, ReactionIcon> = ChatTheme.reactionIconFactory.createReactionIcons(),
+        @DrawableRes showMoreReactionsIcon: Int = R.drawable.stream_compose_ic_more,
+        onDismiss: () -> Unit = {},
+        headerContent: @Composable ColumnScope.() -> Unit = {
+            // Default content, uses own capabilities to regulate sending reactions
+        },
+        centerContent: @Composable ColumnScope.() -> Unit = {
+            // Default content
+        },
+)
+```
+
+The file containing `SelectedMessageMenu` can be
```

Comment:
Also here - feels unfinished and disjointed from the rest

Reply:
Moved the link to the source file and added a more organic transition.